### ISSUE TITLE
kube-proxy failed to clear conntrack for SCTP for the pod which is removed from service

### ISSUE THREAD

**Issue by kyrieSun-wow** [kubernetes/kubernetes#101968]
kube-proxy failed to clear conntrack for SCTP for the pod which is removed from service

#### What happened:
Pod A stop replying the readinessProbe, the corresponding endpoint was removed from service, and so was the corresponding iptables rules, but the conntrack of pod A is stay still. At that time, client sent the SCTP heartbeat to this service via externalIp, this packet still be forwarded to pod A instead of  pod B which is healthy. 

#### What you expected to happen:
After pod A was removed from service, all subsequent packet should be handed over to pod B.(Or am I misunderstand how k8s works ?)

#### Environment:
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:23:04Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}

**Comment by Marcus1911** [comment]
@kyrieSun-wow As a troublesooting, did you try to reproduce that same scenario with just one subflow?

**Comment by kyrieSun-wow** [comment]
> @kyrieSun-wow As a troublesooting, did you try to reproduce that same scenario with just one subflow?

Do you mean just one SCTP association? If so,yes,there is only one SCTP association in my cluster,and this issue recurs 100%.

**Comment by Marcus1911** [comment]
If you could share your repo with docker image and configmap, I can try to reproduce the error and so we can check version and dive in the problem.  
Also reading another issues, it sounds that SCTP has increased the issues amount last weeks.

**Comment by kyrieSun-wow** [comment]
> If you could share your repo with docker image and configmap, I can try to reproduce the error and so we can check version and dive in the problem.
> Also reading another issues, it sounds that SCTP has increased the issues amount last weeks.

Ah , sry,my repo and manifest is related to the business of my company, I can not share them with you,but for reproducing  this issue, any SCTP association will work, you just need to remove the pod from the service by not responding to the ReadInessProbe.

**Comment by aojea** [comment]
This was fixed by @uablrek https://github.com/kubernetes/kubernetes/issues/70020 in 1.19 :thinking: 
Lars you are much into SCTP than me, do you think that current cleanup  is not enough and we should also flush the entries here?
https://github.com/kubernetes/kubernetes/blob/e8cf412e5ec70081477ad6f126c7f7ef7449109c/pkg/proxy/iptables/proxier.go#L1669-L1680

**Comment by uablrek** [comment]
It looks like it would be needed. I actually did not test the previous fix myself, but the SCTP application developers discovering the problem did. I have not heard of any problems since.

So, it is not so easy for me to test (for triage) and fix. I will try to get time this week. But is actually seem not so hard to fix if it only is the code pointed out above anf if you have an SCTP test-env setup, so perhaps a good first issue?

I take it for now, but I might unassign if I can't find time to investigate.

/assign

**Comment by uablrek** [comment]
I thought the important thing was to test if a connection is still active when the service endpoint is removed so I tried a different approach;

1. Connect via SCTP through a service
2. Remove that service
3. Check if the connection is still alive

**And it is! But it is the same for UDP and TCP!**

The problem only exist for `proxy-mode=iptables`, so @kyrieSun-wow, please try `proxy-mode=ipvs` if possible as a work-around.

So to sum up; An active tcp/udp/sctp connection through a service with `proxy-mode=iptables` remains alive even after the service has been deleted (or the endpoint is removed I suppose).

Please take a look at this @aojea @thockin 

/triage accepted

**Comment by uablrek** [comment]
> After pod A was removed from service, all subsequent packet should be handed over to pod B.(Or am I misunderstand how k8s works ?)

For UDP, yes. But TCP and SCTP have connections, so you would have to re-connect and then be connected to B.

**Comment by aojea** [comment]
> So to sum up; An active tcp/udp/sctp connection through a service with `proxy-mode=iptables` remains alive even after the service has been deleted (or the endpoint is removed I suppose).

the connections, in iptables mode,  are simply natted, conntrack is only an "observer", it doesn't interact with the connections, the endpoints (client, server) are the ones in charge of the connection.. 
do you mean that something has to reset the connection when the service is deleted?

**Comment by uablrek** [comment]
> do you mean that something has to reset the connection when the service is deleted?

Yes. Below the loadBalancerIP is `10.0.0.79:7009`, on the node where traffic enters a conntrack entry is created;
```
tcp      6 86396 ESTABLISHED src=192.168.1.201 dst=10.0.0.79 sport=57040 dport=7009 src=11.0.3.2 dst=192.168.1.2 sport=7009 dport=64568 [ASSURED] mark=0 use=1
```
I think this is actually where the NAT state is stored, so the conntrack _is_ involved. When the service is removed the iptables rules are removed but the conntrack entry remains. And it is sufficient for keeping the connection alive.

And this it very easy to test by anyone since it is the same for TCP. Just create some service that you can connect to using telnet or netcat (nc) for instance. Then connect, and remove the service. I have used an external loadBalancerIP but clusterIP seems to be the same.

**Comment by aojea** [comment]
> I think this is actually where the NAT state is stored, so the conntrack is involved. When the service is removed the iptables rules are removed but the conntrack entry remains. And it is sufficient for keeping the connection alive

right , NAT uses the conntrack table, but if we delete the conntrack entry we don't let the TCP connections to end gracefully and both extremes have to timeout. Once conntrack "sees" the close handshake it removes the entry ...

We've seen similar reports here https://github.com/kubernetes/kubernetes/issues/101607 and here https://github.com/kubernetes/kubernetes/issues/100698 , check the latest one as it has a more detailed analysis.
Talking about TCP only, current implementations covers all the general cases and I'd not try to change anything because those problems are hard to debug and hard to spot, and so far iptables is very stable.

**Comment by uablrek** [comment]
https://github.com/kubernetes/kubernetes/issues/101607 is different since it about connections setup, but https://github.com/kubernetes/kubernetes/issues/100698 is the same problem.

There is a difference between terminating the pod and remove the endpoint from the service while the POD is still living (but not ready as in this issue). In the first case I would agree that the conntrack entry should remain, the POD is terminating anyway, but in the second case it is not obvious.

In the `ipvs` case graceful termination (of PODs) is handled differently, by setting weight=0 for the endpoint. But removing the endpoint will rudely black-hole the communication. There are no RST or FIN packets received by the client (or server?). I don't know if that is better for applications with this problem.

So if an endpoint with active connections is removed from a service while the actual server POD lives on (ready or not) we have two scenarios;

**proxy-mode=iptables** - The connection lives on as if nothing has happened. It is up to the application to detect the problem.

**proxy-mode=ipvs** - The connection is black-holed. Communication just stops. Application can not count on a connection close or reset but must detect the problem with keep-alive or timeouts.

It doesn't seem to be a big problem since it must have been like this since day one and only a few issues.

How to proceed? Just document the facts?

**Comment by uablrek** [comment]
Here is an example on the client-side (external) with `ipvs` when the service is removed;
```
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp6       0     20 1000::1:c0a8:1c9:46804  1000::79:7009           ESTABLISHED 390/ncat            
```
Data will accumulate in the `Send-Q` until the connection is terminated by the client.

**Comment by aojea** [comment]
> How to proceed? Just document the facts?

great explanation, I'm leaning towards documentation , @thockin this is a good reading ^^^ can you weigh in?

**Comment by uablrek** [comment]
@thockin Please comment on https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-843272392

**Comment by khenidak** [comment]
> @thockin Please comment on [#101968 (comment)](https://github.com/kubernetes/kubernetes/issues/101968#issuecomment-843272392)

@uablrek spot on. if the client has a session TCP or SCTP What will happen is:
1. whatever node they are connected to will detect endpoint out of set. Then it removes iptables entries + clear iptables.
2. new packets coming in will be dropped (we don't send ICMP destination unreachable iirc for cstate==INVALID). 
3. client - because it is a stateful connection - will go try to resend and eventually times out and terminates the connection. e.g., There will be no acks coming back from server.
4. Client code should then re-attempt to connect which creates a new entry in conntrack and new packets of that connection will just work as expected.

If the client does not know how to terminate and re-connect then the client will go into `hang: waiting for response` (ugly, i know) depending on how the client configured this can be milliseconds, seconds or even minutes, specially for long running sessions. This is even worse when there is an external LB in the middle that blindly sends keep-alives.

I am +1 on @aojea  to provide doc and maybe an example and proposed config changes to short-circuit the fail-over.
